lib/Wikia/src/Service/User/Attributes/UserAttributes.php

```diff
@@ -91,9 +97,50 @@ private function setInMemcache( $userId, $attributes ) {
 	 * @param Attribute $attribute
 	 */
 	public function setAttribute( $userId, Attribute $attribute ) {
-		$this->setInService( $userId, $attribute );
+		$this->loadAttributes( $userId );
+
+		// If attribute value is null and default exists, set value to default
+		if ( is_null( $attribute->getValue() ) && isset( $this->defaultAttributes[$attribute->getName()] ) ) {
+			$attribute->setValue( $this->defaultAttributes[$attribute->getName()] );
+		}
+
 		$this->setInInstanceCache( $userId, $attribute );
-		$this->setInMemcache( $userId, $this->attributes[$userId] );
+	}
+
+	public function save( $userId ) {
+		$attributes = $this->loadAttributes( $userId );
+
+		// TODO When bulk updates are complete, convert this to a single request
+		$savedAttributes = [];
+		foreach( $attributes as $name => $value ) {
+			if ( $this->attributeShouldBeSaved( $name, $value ) ) {
+				$this->setInService( $userId, new Attribute( $name, $value ) );
+				$savedAttributes[$name] = $value;
+			} elseif ( $this->attributeShouldBeDeleted( $name, $value ) ) {
+				$this->deleteFromService( $userId, new Attribute( $name, $value ) );
+			}
+		}
+
+		$this->setInMemcache( $userId, $savedAttributes );
+	}
+
+	/**
+	 * Returns true if either is true:
+	 * 1.) No default for the attribute and the value is either not false or not null
+	 * 2.) Value is different than the default
+	 * @param $name
+	 * @param $value
+	 * @return bool
+	 */
+	private function attributeShouldBeSaved( $name, $value ) {
+		return (
+			( is_null( $this->defaultAttributes[$name] ) ) && !( $value === false || is_null( $value ) ) ||
+			$value != $this->defaultAttributes[$name]
+		);
+	}
+
+	private function attributeShouldBeDeleted( $name, $value ) {
+		return isset( $this->defaultAttributes[$name] ) && $value == $this->defaultAttributes[$name];
```

Comment:
I don't think so. Only because that's not the original behavior MW had.

When an attribute was set to a null value, it would [try and set the defaultValue](https://github.com/Wikia/app/blob/dev/includes/User.php#L2882-L2884) if it exists. But if no default existed, then it would be set as null.

Then when it came time to save, it would [check if an attribute had a default value](https://github.com/Wikia/app/blob/dev/includes/User.php#L5012-L5015) and delete it if so. If it's null without a default howefver, it would still get set in the db. So there may be places in MW where it expects/uses these null values